src/rtpi/condition_variable.hpp

```diff
@@ -0,0 +1,227 @@
+	{
+		using duration = std::chrono::steady_clock::duration;
+		auto relative_time =
+			std::chrono::duration_cast<duration>(rel_time);
+		if (relative_time < rel_time)
+			++relative_time;
+		return wait_until(lock, std::chrono::steady_clock::now() +
+						relative_time);
+	}
+
+	// Overload that takes a predicate. This overload may be used to
+	// ignore spurious awakenings while waiting for a specific condition
+	// to become true.
+	template <class Clock, class Period, class Predicate>
+	bool wait_for(std::unique_lock<rtpi::mutex> &lock,
+		      const std::chrono::duration<Clock, Period> &rel_time,
+		      Predicate stop_waiting)
+	{
+		using duration = std::chrono::steady_clock::duration;
+		auto relative_time =
+			std::chrono::duration_cast<duration>(rel_time);
+		if (relative_time < rel_time)
+			++relative_time;
+		return wait_until(
+			lock, std::chrono::steady_clock::now() + relative_time,
+			std::move(stop_waiting));
+	}
+
+	// Atomically releases lock, blocks the current executing thread, and
+	// adds it to the list of threads waiting on *this. The thread will be
+	// unblocked when notify_all() or notify_one() is executed, or when
+	// the absolute time point timeout_time is reached. It may also be
+	// unblocked spuriously. When unblocked, regardless of the reason,
+	// lock is reacquired and wait_until exits.
+	template <class Duration>
+	cv_status
+	wait_until(std::unique_lock<rtpi::mutex> &lock,
+		   const std::chrono::time_point<std::chrono::steady_clock,
+						 Duration> &timeout_time)
+	{
+		return wait_until_impl(lock, timeout_time);
+	}
+
+	template <class Clock, class Duration>
+	cv_status
+	wait_until(std::unique_lock<rtpi::mutex> &lock,
+		   const std::chrono::time_point<Clock, Duration> &timeout_time)
+	{
+		using std::chrono::steady_clock;
+
+		const typename Clock::time_point user_clock_entry =
+			Clock::now();
+		const typename steady_clock::time_point steady_clock_entry =
+			steady_clock::now();
+
+		const auto delta = timeout_time - user_clock_entry;
+		const auto steady_timeout_time = steady_clock_entry + delta;
+
+		if (wait_until_impl(lock, steady_timeout_time) ==
+		    cv_status::no_timeout) {
+			return cv_status::no_timeout;
+		}
+
+		// We got a timeout against CLOCK_MONOTONIC but we need to check
+		// against the caller-supplied clock.
+		if (Clock::now() < timeout_time) {
+			return cv_status::no_timeout;
+		}
+
+		return cv_status::timeout;
+	}
+
+	// Overload that takes a predicate. This overload may be used to
+	// ignore spurious awakenings while waiting for a specific condition
+	// to become true.
+	template <class Clock, class Duration, class Predicate>
+	bool
+	wait_until(std::unique_lock<rtpi::mutex> &lock,
+		   const std::chrono::time_point<Clock, Duration> &timeout_time,
+		   Predicate stop_waiting)
+	{
+		while (!stop_waiting) {
+			if (wait_until(lock, timeout_time) ==
+			    cv_status::timeout)
+				return stop_waiting();
+		}
+	}
```

Comment:
Missing return value^?

Reply:
hah, and this is how I learn that libstdc++'s tests never try the `Predicate`-taking `wait_until` overload...